### PR TITLE
[gold] Enable time trace profiler in LLVMgold

### PR DESCRIPTION
**Pull request by mshockwave**
To get the time trace of LTO in the gold plugin, now we can pass the `time-trace=<time trace file>` as well as
`time-trace-granularity=<granularity>` flags to LLVMgold.

Note that we still have to populate `LTOConfig::TimeTraceEnabled` and `LTOConfig::TimeTraceGranularity` because ThinLTO backend needs them.

### CODE REVIEW

File: llvm/tools/gold/gold-plugin.cpp [hunk 1]
```diff
@@ -1113,6 +1131,19 @@ static ld_plugin_status allSymbolsReadHook() {
   if (unsigned NumOpts = options::extra.size())
     cl::ParseCommandLineOptions(NumOpts, &options::extra[0]);
 
+  // Initialize time trace profiler
+  if (!options::time_trace_file.empty())
+    llvm::timeTraceProfilerInitialize(options::time_trace_granularity,
+                                      options::extra.size() ? options::extra[0]
+                                                            : "LLVMgold");
```
Comment (teresajohnson) [inline llvm/tools/gold/gold-plugin.cpp:1138]: At https://github.com/llvm/llvm-project/blob/6cd86d0fae8cbb752a713860f131b9b759b2cbb8/llvm/tools/gold/gold-plugin.cpp#L316 we set extra[0] to "LLVMgold" if extra is empty, so this shouldn't be needed (extra should never be empty and extra[0] can always be used).

**Review by teresajohnson — APPROVED** [review]
lgtm with a minor change below

File: llvm/tools/gold/gold-plugin.cpp (hunk 1 above)
Comment (mshockwave) [reply llvm/tools/gold/gold-plugin.cpp:1138]: I think we reach this line (line 316) only if there are additional `-plugin-opt` that LLVMgold doesn't recognize. If there aren't any of those, I think `options::extra` will remain empty.

File: llvm/test/tools/gold/X86/time-trace.ll [hunk 2]
```diff
@@ -0,0 +1,32 @@
+; RUN: llvm-as %s -o %t.o
+
+; RUN: %gold -plugin %llvmshlibdir/LLVMgold%shlibext \
+; RUN:    -m elf_x86_64 --plugin-opt=time-trace=%t2.json \
+; RUN:    -shared %t.o -o %t3.s
```
Comment (MaskRay) [inline llvm/test/tools/gold/X86/time-trace.ll:5]: If the output is unused, prefer `-o /dev/null`; otherwise `.so` for shared objects.

**Review by MaskRay — APPROVED** [review]

File: llvm/tools/gold/gold-plugin.cpp (hunk 1 above)
Comment (teresajohnson) [reply llvm/tools/gold/gold-plugin.cpp:1138]: got it thanks

**Comment by tstellar** [comment]
Hi @mshockwave we are seeing test failures on s390x with this change:
https://download.copr.fedorainfracloud.org/results/@fedora-llvm-team/llvm-snapshots-big-merge-20240607/fedora-39-s390x/07546351-llvm/builder-live.log.gz

File: llvm/tools/gold/gold-plugin.cpp [hunk 3]
```diff
@@ -308,6 +314,14 @@ namespace options {
       RemarksFormat = std::string(opt);
     } else if (opt.consume_front("stats-file=")) {
       stats_file = std::string(opt);
+    } else if (opt.consume_front("time-trace=")) {
+      time_trace_file = std::string(opt);
+    } else if (opt.consume_front("time-trace-granularity=")) {
+      unsigned Granularity;
+      if (opt.getAsInteger(10, Granularity))
+        message(LDPL_FATAL, "Invalid time trace granularity: %s", opt);
```
Comment (nikic) [inline llvm/tools/gold/gold-plugin.cpp:322]: This is passing a StringRef to a varargs function looking for `const char *`. This is obviously wrong, but it looks like the ABI happens to work out fine on non-s390x architectures...

Comment (nikic) [reply llvm/tools/gold/gold-plugin.cpp:322]: I've put up https://github.com/llvm/llvm-project/pull/95083 to fix this.

**Comment by mshockwave** [comment]
> Hi @mshockwave we are seeing test failures on s390x with this change: https://download.copr.fedorainfracloud.org/results/@fedora-llvm-team/llvm-snapshots-big-merge-20240607/fedora-39-s390x/07546351-llvm/builder-live.log.gz

Sorry the notification must slipped thru my mailbox during the weekend. Thank you @nikic for the quick fix!

**Pull request merged**